cocos/audio/android/UrlAudioPlayer.cpp

```diff
@@ -0,0 +1,339 @@
+/****************************************************************************
+Copyright (c) 2016 Chukong Technologies Inc.
+
+http://www.cocos2d-x.org
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+
+#define LOG_TAG "UrlAudioPlayer"
+
+#include "audio/android/UrlAudioPlayer.h"
+#include "audio/android/ICallerThreadUtils.h"
+#include "audio/android/AssetFd.h"
+
+#include <math.h>
+#include <algorithm> // for std::find
+
+namespace cocos2d { namespace experimental {
+
+static std::vector<UrlAudioPlayer *> __allPlayers;
+
+static std::once_flag __onceFlag;
+static int __instanceCount = 0;
+
+UrlAudioPlayer::UrlAudioPlayer(SLEngineItf engineItf, SLObjectItf outputMixObject, ICallerThreadUtils* callerThreadUtils)
+        : _engineItf(engineItf), _outputMixObj(outputMixObject),
+          _callerThreadUtils(callerThreadUtils), _id(-1), _assetFd(nullptr),
+          _playObj(nullptr), _playItf(nullptr), _seekItf(nullptr), _volumeItf(nullptr),
+          _volume(0.0f), _duration(0.0f), _isLoop(false), _state(State::INVALID),
+          _isDestroyed(false), _playEventCallback(nullptr)
+{
+    std::call_once(__onceFlag, []() {
+        ALOGV("Initializing static variables in UrlAudioPlayer ...");
+        __allPlayers.reserve(16);
+    });
+    ++__instanceCount;
+    ALOGV("Current UrlAudioPlayer instance count: %d", __instanceCount);
+    __allPlayers.push_back(this);
+}
+
+UrlAudioPlayer::~UrlAudioPlayer()
+{
+    ALOGV("~UrlAudioPlayer(): %p", this);
+    --__instanceCount;
+    auto iter = std::find(__allPlayers.begin(), __allPlayers.end(), this);
+    if (iter != __allPlayers.end())
+    {
```

Comment:
Yep, if the instance of UrlAudioPlayer is destroyed, the counter should be decreased.
It will be easier for checking whether there is leak.